logstash-core/src/main/java/org/logstash/common/LsQueueUtils.java

```diff
@@ -78,17 +87,24 @@ private static int drain(final BlockingQueue<JrubyEventExtLibrary.RubyEvent> que
         final Collection<JrubyEventExtLibrary.RubyEvent> collection, final int count,
         final long nanos) throws InterruptedException {
         final long deadline = System.nanoTime() + nanos;
+        long pollTimeout = nanos;
         int added = 0;
         do {
             added += queue.drainTo(collection, count - added);
+            
             if (added < count) {
+                if (pollTimeout <= 0) {
```

Comment:
@colinsurprenant this is without effect and just adds dead code + is slower. A negative timeout value on a `poll` makes it return without blocking => much better to make the non-blocking `poll` to find out that the timeout ran out than to `break` out and miss the chance to get an `Event` without blocking

Reply:
well, the behaviour for polling with a negative timeout is undefined at the API level, in the implementation (which could change) it will first try drain the queue regardless of the timeout value and then exit on <= 0 timeout. So if we want to be correct WRT the timeout then we should not call poll() when timeout is exhausted.